File: Day6/finding_pairs_with_a_certain_sum.js
class FindSumPairs {
    constructor(nums1, nums2) {
        this.nums1 = [...nums1];
        this.nums2 = [...nums2];
        this.freqMap = new Map();
        for (let num of nums2) {
            this.freqMap.set(num, (this.freqMap.get(num) || 0) + 1);
        }
    }

    add(index, val) {
        const prevVal = this.nums2[index];
        this.nums2[index] += val;

        this.freqMap.set(prevVal, this.freqMap.get(prevVal) - 1);
        if (this.freqMap.get(prevVal) === 0) {
            this.freqMap.delete(prevVal);
        }

        const newVal = this.nums2[index];
        this.freqMap.set(newVal, (this.freqMap.get(newVal) || 0) + 1);
    }

    count(tot) {
        let cnt = 0;
        for (let num1 of this.nums1) {
            const num2 = tot - num1;
            cnt += this.freqMap.get(num2) || 0;
        }
        return cnt;
    }
}

File: Day6/finding_pairs_with_a_certain_sum.py
class FindSumPairs:

    def __init__(self, nums1: List[int], nums2: List[int]):
        self.nums1 = nums1[:]
        self.nums2 = nums2[:]
        self.freqMap = Counter(nums2)

    def add(self, index: int, val: int) -> None:
        prevVal = self.nums2[index]
        self.nums2[index] += val
        self.freqMap[prevVal] -= 1
        if self.freqMap[prevVal] == 0:
            del self.freqMap[prevVal]
        self.freqMap[self.nums2[index]] += 1

    def count(self, tot: int) -> int:
        cnt = 0
        for num1 in self.nums1:
            num2 = tot - num1
            cnt += self.freqMap.get(num2, 0)
        return cnt
